### PR TITLE
drivers/lsm6dsox: Add support for SPI mode.

### PR DESCRIPTION
**Pull request by iabdalkader**
Note: Tested both the basic and MLC examples in I2C and SPI modes.

### CODE REVIEW

File: drivers/lsm6dsox/lsm6dsox.py [hunk 1]
```diff
@@ -4,8 +4,8 @@
 
 The MIT License (MIT)
 
-Copyright (c) 2021 Damien P. George
-Copyright (c) 2021 Ibrahim Abdelkader <iabdalkader@openmv.io>
+Copyright (c) 2022 Damien P. George
+Copyright (c) 2022 Ibrahim Abdelkader <iabdalkader@openmv.io>
```
Comment (dpgeorge) [inline drivers/lsm6dsox/lsm6dsox.py:8]: This file was created in 2021, so the copyright should stay at that for me.  For yourself, you can change it to `2021-2022`.

File: drivers/lsm6dsox/lsm6dsox.py [hunk 2]
```diff
@@ -30,16 +30,24 @@
...
     print('Accelerometer: x:{:>8.3f} y:{:>8.3f} z:{:>8.3f}'.format(*lsm.read_accel()))
     print('Gyroscope:     x:{:>8.3f} y:{:>8.3f} z:{:>8.3f}'.format(*lsm.read_gyro()))
     print("")
     time.sleep_ms(100)
 """
+
 import array
+from machine import I2C
+from machine import SPI
+from machine import Pin
```
Comment (dpgeorge) [inline drivers/lsm6dsox/lsm6dsox.py:50]: I think it's best if this driver (or any driver for that matter) does not import hardware classes like this.  It makes the driver much more general if it doesn't rely on specific hardware in any way.

See further comments below how to do it.

File: drivers/lsm6dsox/lsm6dsox.py [hunk 3]
```diff
@@ -88,10 +97,15 @@ def __init__(
         accel_scale: (+/-2g, +/-4g, +/-8g, +-16g)
         ucf: MLC program to load.
         """
-        self.i2c = i2c
+        self.bus = bus
+        self.cs_pin = cs_pin
         self.address = address
 
+        if cs_pin is None and isinstance(self.bus, SPI):
```
Comment (dpgeorge) [inline drivers/lsm6dsox/lsm6dsox.py:104]: I suggest to check for attributes to determine the bus kind.  Eg:
```python
self._use_i2c = hasattr(self.bus, "readfrom_mem")
if cs_pin is None and not self._use_i2c:
    raise ValueError(...)
```

File: drivers/lsm6dsox/lsm6dsox.py [hunk 4]
```diff
@@ -88,10 +97,15 @@ def __init__(
...
+            raise ValueError("Must provide CS pin in SPI mode.")
+
         # check the id of the Accelerometer/Gyro
+        print("0x%x" % (self.__read_reg(_WHO_AM_I_REG)))
```
Comment (dpgeorge) [inline drivers/lsm6dsox/lsm6dsox.py:108]: I think this is a stay debug print that can be removed.

File: drivers/lsm6dsox/lsm6dsox.py [hunk 5]
```diff
@@ -149,13 +163,39 @@ def __init__(
         self.accel_scale = 32768 / accel_scale
 
     def __read_reg(self, reg, size=1):
-        buf = self.i2c.readfrom_mem(self.address, reg, size)
+        if isinstance(self.bus, SPI):
```
Comment (dpgeorge) [inline drivers/lsm6dsox/lsm6dsox.py:166]: As above, this would be `if self._use_i2c:`.

File: drivers/lsm6dsox/lsm6dsox.py (hunk 3 above)
Comment (iabdalkader) [reply drivers/lsm6dsox/lsm6dsox.py:104]: It's not likely, but what if another bus/obj has that attr ? I could just check if `cs_pin` is set or not?

Comment (iabdalkader) [reply drivers/lsm6dsox/lsm6dsox.py:104]: Using `readfrom_mem` for now.

Comment (dpgeorge) [reply drivers/lsm6dsox/lsm6dsox.py]: The idea for using hasattr like this is that you can provide any class as the bus, so long as it has the correct methods.  So strictly you want something like:
```python
if hasattr(bus, "read") and hasattr(bus, "write"):
   self._bus_kind = "SPI"
elif hasattr(bus, "readfrom_mem") and hasattr(bus, "writeto_mem"):
    self._bus_kind = "I2C"
```
But I just tried to make it a simpler version of that, and I think it's enough to just check for `readfrom_mem`.  If you require either an I2C-like instance or SPI-like instance, checking for `readfrom_mem` should be enough to distinguish them.

Comment (dpgeorge) [reply drivers/lsm6dsox/lsm6dsox.py]: But, yes, the simpler alternative is to just check for `cs_pin` being None or not, and use that to decide if it's SPI or I2C.  That might actually be better, then you can just use `if cs_pin: ...` in the register helper methods.

Comment (iabdalkader) [reply drivers/lsm6dsox/lsm6dsox.py]: I actually went ahead and implemented your first suggestion, after thinking about it the user may forget to set the cs_pin and still want SPI. However, let me know if you want to switch to checking `cs_pin` either way is fine by me.

Comment (dpgeorge) [reply drivers/lsm6dsox/lsm6dsox.py]: > Using readfrom_mem for now.

I do like the idea of just using `cs_pin`.  That is the simplest option, and means you don't need to raise an exception if `cs_pin` is not given but the bus is detected as SPI.

Comment (iabdalkader) [reply drivers/lsm6dsox/lsm6dsox.py]: I could add this:

```Python
if hasattr(bus, "read") and hasattr(bus, "write"):
   self._bus_kind = "SPI"
elif hasattr(bus, "readfrom_mem") and hasattr(bus, "writeto_mem"):
    self._bus_kind = "I2C"
else:
    raise ValueError("Unsupported bus")

if self._bus_kind == 'SPI' and self.cs_pin is None:
    raise ValueError("A CS pin must be provided in SPI mode.")
```

Comment (dpgeorge) [reply drivers/lsm6dsox/lsm6dsox.py]: That's overkill, IMO.

You can just leave it as-is with the `readfrom_mem` used to detect.

Comment (iabdalkader) [reply drivers/lsm6dsox/lsm6dsox.py]: Okay, just updated the exception message.

**Comment by dpgeorge** [comment]
Merged in 364569d25f9db4ab8f170e20489daaebe36502a1

**Pull request closed without merging**